dune/gdt/operators/interfaces.hh

```diff
@@ -181,807 +385,776 @@ public:
-    ret.add(*this, 1.);
-    ret.add(other, 1.);
-    return ret;
-  }
+    DUNE_THROW(XT::Common::Exceptions::you_are_using_this_wrong,
+               "Given type is not one of jacobian_options()!"
+                   << "\n"
+                   << "   type = " << type << "\n"
+                   << "   jacobian_options() = " << print(this->jacobian_options()));
+    return XT::Common::Configuration();
+  } // ... jacobian_options(...)
 
-  virtual LincombOperatorType operator+(ThisType& other)
+protected:
+  /**
+   * Used by derived DiscreteOperators to feed the implementations of jacobian_options().
+   * Simply return a vector of configurations with each at least a "type" key.
+   */
+  virtual std::vector<XT::Common::Configuration> all_jacobian_options() const
   {
-    std::string derived_logging_prefix = "";
-    if (this->logger.debug_enabled) {
-      derived_logging_prefix = "LincombOperator";
-      this->logger.debug() << "operator+(other_op=" << &other << ")" << std::endl;
-    }
-    LincombOperatorType ret(this->source_space(), this->range_space(), derived_logging_prefix);
-    ret.add(*this, 1.);
-    ret.add(other, 1.);
-    return ret;
+    return {};
   }
 
+public:
   /**
-   * \note vector is interpreted as a ConstantOperator
-   * \sa ConstantOperator
-   */
-  virtual LincombOperatorType operator+(const VectorType& vector)
+   * \brief Either appends suitable functors to the jacobian_op (such that the jacobian of this operator is assembled
+   *        additively into jacobian_op) or adds the jacobian of this operator to jacobian_op.matrix().
+   *
+   * \note You need to call jacobian_op.assemble() to be sure to have the jacobian fully assembled.
+   **/
+  virtual void jacobian(const VectorType& source_vector,
+                        MatrixOperatorType& jacobian_op,
+                        const XT::Common::Configuration& opts,
+                        const XT::Common::Parameter& param = {}) const
   {
-    std::string derived_logging_prefix_lop = "";
-    std::string derived_logging_prefix_cop = "";
-    if (this->logger.debug_enabled) {
-      derived_logging_prefix_lop = "LincombOperator";
-      derived_logging_prefix_cop = "ConstantOperator";
-      this->logger.debug() << "operator+(vector.sup_norm()=" << vector.sup_norm() << ")" << std::endl;
+    LOG_(debug) << "jacobian(source_vector.sup_norm()=" << source_vector.sup_norm()
+                << ", jacobian_op.matrix().sup_norm()=" << jacobian_op.matrix().sup_norm()
+                << ",\n   opts=" << print(opts, {{"oneline", "true"}}) << ",\n   param=" << print(param) << ")"
+                << std::endl;
+    this->assert_matching_source(source_vector);
+    DUNE_THROW_IF(!opts.has_key("type"), Exceptions::operator_error, "opts = " << opts);
+    const auto type = opts.get<std::string>("type");
+    if (type == "zero") {
+      LOG_(info) << "not adding zero jacobian ..." << std::endl;
```

Comment:
Allow adding zero operators here, but throwing an error otherwise is intentional? If there's a structural reason for this, maybe add a note?

Reply:
This is a default implementation in the interface for convenience. Ops with zero jacobian only need to implement `available...`, not `jacobian()`. So yes, some docs should be there.